gobblin-runtime/src/main/java/gobblin/runtime/MysqlDatasetStateStore.java

```diff
@@ -51,9 +70,20 @@
 
   private static final Logger LOGGER = LoggerFactory.getLogger(MysqlDatasetStateStore.class);
 
+  private static final String SELECT_JOB_STATE_LATEST_BY_PREFIX_TEMPLATE =
+          "SELECT state, CASE WHEN locate(?, table_name) = 0 THEN '' "
```

Comment:
I think we can change the `MysqlStateStore.SELECT_JOB_STATE_LATEST_TEMPLATE` query to: 
``` sql
SELECT state 
FROM $TABLE$ WHERE store_name = ?
ORDER BY CAST(REVERSE(SUBSTRING_INDEX(REVERSE(table_name), '_', 1)) AS UNSIGNED) DESC
LIMIT 1
```

This has the benefit of just taking the segment of the table name that appears after the last `_`.  While this may include an extension, the `CAST` correctly ignores it. Then we can order by the sequence number and take the latest.

As far as performance, running this query against my job history db on a task with 168k entries in the DB, I'm able to get a result in 400ms.  I think that is adequate performance for this scenario as this is not called in a tight loop and other tasks will dominate the processing time.

Now, as for the `MysqlDatasetStateStore.SELECT_JOB_STATE_LATEST_BY_PREFIX_TEMPLATE`, I think we can do something like the following:

``` sql
SELECT
    state,
    CASE WHEN locate(?, table_name) = 0 THEN '' ELSE SUBSTRING_INDEX(table_name, ?, 1) END AS urn
FROM $TABLE$ t1
INNER JOIN (
    SELECT
        CASE WHEN locate(?, table_name) = 0 THEN '' ELSE SUBSTRING_INDEX(table_name, ?, 1) END AS urn,
        MAX(CAST(REVERSE(SUBSTRING_INDEX(REVERSE(table_name), ?, 1)) AS UNSIGNED)) AS sequence
    FROM $TABLE$
    WHERE store_name = ?
    GROUP BY urn
) t2 ON CASE WHEN locate(?, table_name) = 0 THEN '' ELSE SUBSTRING_INDEX(table_name, ?, 1) END = t2.urn
     AND CAST(REVERSE(SUBSTRING_INDEX(REVERSE(table_name), ?, 1)) AS UNSIGNED) = t2.sequence
WHERE store_name = ?
```

This has a similar behavior to the above query.